### PR TITLE
support for dynamic variables in template context (multi protocol execution)

### LINKED ISSUE #3413: Global / Dynamic variables to support in template context  (multi protocol execution)
### Please describe your feature request:
From discord -

> Is there a way to use extracted values from requests in the dns section? I have extracted a value after making a web request; however, it doesn't seem available from the subsequent dns section.

### Current behavior:

Extracted variables are only available within the same request context.

### Desired behavior

Extracted variables should be accessible within the template context to any or all protocols.

### Dynamic extractor to use with multi-protocol template

```yaml
id: dns-http-template

info:
  name: dns + http takeover template
  author: pdteam
  severity: info

dns:
  - name: "{{FQDN}}" # DNS Request
    type: cname

    extractors:
      - type: dsl
        name: cname
        dsl:
          - cname
        internal: true

http:
  - method: GET # http request
    path:
      - "{{BaseURL}}"

    matchers:
      - type: dsl
        dsl:
          - contains(body,'Domain not found') # check for http string
          - contains(cname, 'github.io') # check for cname (extracted information from dns response)
        condition: and
```

### Global protocol variables with multi-protocol template

```yaml
id: nuclei-multi-protocol

info:
  name: multi protocol support
  author: pdteam
  severity: info

dns:
  - name: "{{FQDN}}" # dns request 
    type: cname

ssl:
  - address: "{{Hostname}}" # ssl request

http:
  - method: GET
    path:
      - "{{BaseURL}}" # http request

    headers:
      Host: "{{ssl_subject_cn}}" # host extracted from ssl request

    matchers:
      - type: dsl
        dsl:
          - contains(http_body,'Domain not found') # check for http string
          - contains(dns_cname, 'github.io') # check for cname
        condition: and
```

### PR DESCRIPTION
**Pull request by tarunKoyalwar**
## Proposed changes

- adds support for dynamic variables in template context (i.e when template contains more than 1 protocol)
- adds support for global variables in template context prefixed by protocol name.
- adds support for multi protocol request execution
- closes #3413 

## Checklist

- [x] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [ ] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by ehsandeep — CHANGES_REQUESTED** [review]
we can also expose protocol level global variables (prefixed by protocol name) in template context to use with other protocols.

example added in https://github.com/projectdiscovery/nuclei/issues/3413

**Comment by Mzack9999** [comment]
Recommendations:
- Consider passing to next steps only named extractors with a unique value as a start, this resolves the issue to prefix the name with the protocol. This drops the need to prefix the protocol to the name.
- Evaluate if the contextargs mechanism might be a suitable transport vector for bringing in/out values from protocol context
- interactsh handling might not be possible (ref. https://github.com/projectdiscovery/nuclei/issues/3437 and https://github.com/projectdiscovery/nuclei/pull/2930#issuecomment-1558575851)

**Comment by tarunKoyalwar** [comment]
### New Changes
- use `contextArgs` to pass and add values to map in template context
- fix race condition in contextArgs
- add design related docs related to multi protocol at https://github.com/projectdiscovery/nuclei/tree/issue-3413-multi-proto-support/v2/pkg/protocols/multi#readme
- sort keys before vardump
- unless template is a multi protocol all logic is no-op (this reduces memory consumption of maps compared to earlier implementation)

### To Discuss / Confirm ?
- in mulit protocol template all protocol response fields are prefixed with protocol ex: `http_status_code` . just in case `status_code` variables is also available in matchers/extractors should that be removed ?? (cc: @ehsandeep , @Mzack9999 )

**Comment by tarunKoyalwar** [comment]
### example complex template

```yaml
id: nuclei-multi-protocol

info:
  name: multi protocol support
  author: pdteam
  severity: info

variables:
  # we can use variables as pre-processing since it is evaluated after each protocol
  # example replace CN= prefix from ssl_subject_dn
  subject_dn_filtered: '{{replace(ssl_subject_dn,"CN=","")}}' # ssl_subject_dn is obtained from ssl
  # subject_dn_filtered is used in http protocol

dns:
  - name: "{{FQDN}}" # dns request 
    type: cname

ssl:
  - address: "{{Hostname}}" # ssl request

http:
  - method: GET
    path:
      - "{{BaseURL}}" # http request

    headers:
      Host: "{{ssl_subject_cn}}" # host extracted from ssl request
      Cipher: "{{ssl_cipher}}"

    matchers:
      - type: dsl
        dsl:
          # - contains(http_body,'File not found') # check for http string
          - http_status_code == 404
          - contains(dns_cname, 'github.io') # check for cname
          - subject_dn_filtered == 'mta-sts.managed.hackerone.com'
        condition: and 
```

```console
$ ./nuclei -u mta-sts.managed.hackerone.com -t ~/test-templates/multiprotocomplex.yaml  -ms

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.9.5-dev

		projectdiscovery.io

[WRN] Found 24 templates loaded with deprecated paths, update before v2.9.5 for continued support.
[INF] Current nuclei version: v2.9.5-dev (development)
[INF] Current nuclei-templates version: v9.5.0 (latest)
[INF] New templates added in latest release: 62
[INF] Templates loaded for current scan: 1
[INF] Targets loaded for current scan: 1
[INF] Running httpx on input host
[INF] Found 1 URL from httpx
[nuclei-multi-protocol] [matched] [http] [info] https://mta-sts.managed.hackerone.com
```

**Comment by tarunKoyalwar** [comment]
- requested review regarding implmentation before proceeding with tests & docs

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
Neat implementation with full reuse of `contextargs` transport vector as it was designed - I'm suggesting some minor changes and a potential new helper for utils.

Note: the smart-input functionality only executes multiprotocol templates if the first in-order protocol matches the input type and blocks at any intermediate step when the input doesn't match the protocol type.

File: v2/pkg/protocols/common/utils/vardump/dump.go [hunk 1]
```diff
@@ -21,7 +23,12 @@ func DumpVariables(data map[string]interface{}) string {
 	buffer.Grow(len(data) * 78) // grow buffer to an approximate size
 
 	builder := &strings.Builder{}
-	for k, v := range data {
+	// sort keys for deterministic output
+	keys := maps.Keys(data)
```
[the remark this answers is not in the document]
Comment (tarunKoyalwar) [reply v2/pkg/protocols/common/utils/vardump/dump.go]: done

**Review by Mzack9999 — APPROVED** [review]
lgtm! - Just suggesting to consider using synclock map from the maputils package as it's safe for concurrent access and with protected versions of various methods

**Comment by tarunKoyalwar** [comment]
@ehsandeep  @ShubhamRasal  will add docs related to multi protocol in nuclei-docs soon . requesting final review regarding  this PR

**Review by ShubhamRasal — APPROVED** [review]
lgtm 🚀 🔥

**Comment by tarunKoyalwar** [comment]
### Documentation

- [ ] https://github.com/projectdiscovery/nuclei-docs/pull/151

**Review by ehsandeep — APPROVED** [review]
@tarunKoyalwar merge conflict ^

**Pull request merged**